### PR TITLE
Realization of hashtable with Linear Probing

### PR DESCRIPTION
**Pull request by gorshkovayulia**
@ctapobep please take a look at the current version (maybe only code). It's not ideal, I'm planning to finish this tomorrow (improve my tests and add necessary info).

### CODE REVIEW

File: src/main/java/hashtable/HashTable2.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package hashtable;
+
+public class HashTable2 {
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:3]: Let's give it a speaking name: LinearProbingHashTable

File: src/main/java/hashtable/HashTable2.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    private Item[] table;
+    private static final double DEFAULT_LOAD_FACTOR = 0.75;
+    private int count;
+
+    public HashTable2(int size) {
+        if(size <= 0) {
+            throw new IllegalArgumentException("The size of table must be more than 0!");
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:10]: more -> greater

File: src/main/java/hashtable/HashTable2.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        table = new Item[size];
+        for (int i = 0; i < size; i++) {
+            table[i] = null;
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:14]: Arrays of objects contain `null`s by default, no need to do this explicitly.

File: src/main/java/hashtable/HashTable2.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+
+    public void put(Object key, Object value) {
+        Item item = new Item(key, value);
+        int idx = findIdxToInsert(key);
+        table[idx] = item;
+        count++;
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:22]: `size` sounds better

File: src/main/java/hashtable/HashTable2.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        if(table[hash] != null && table[hash] != Item.DELETED) {
+            table[hash] = Item.DELETED;
+        }
+    }
+
+    public Object get(Object key) {
+        if (key == null) {
+            return null;
+        }
+        int hash = findIdxToInsert(key);
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:40]: It's not a hash, it's an index

File: src/main/java/hashtable/HashTable2.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        if (table[hash] == null || table[hash] == Item.DELETED) {
+            return 0;
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:42]: return null

File: src/main/java/hashtable/HashTable2.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        } else {
+            return table[hash].getValue();
+        }
+    }
+
+    int getSize() {
+        return table.length;
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:49]: did you mean to return `count` here?

File: src/main/java/hashtable/HashTable2.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    Item getContent(int index) {
+        return table[index];
+    }
+
+    String printSlot(int index) {
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:56]: Let's try testing without this method this time

File: src/main/java/hashtable/HashTable2.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        return table[index] == null ? "null" : table[index].printContent();
+    }
+
+    private int findIdxToInsert(Object key) {
+        if (key == null) {
+            throw new IllegalArgumentException("Key cannot be null!");
+        }
+        int hash = Math.abs(key.hashCode() % table.length);
+        while (table[hash] != null && table[hash].getKey() != key && table[hash] != Item.DELETED)
```
Comment (ctapobep) [inline src/main/java/hashtable/HashTable2.java:65]: Did you mean to compare keys by reference? Or by equals()?

File: src/main/java/hashtable/HashTable2.java (hunk 7 above)
Comment (gorshkovayulia) [reply src/main/java/hashtable/HashTable2.java:49]: No, I need the length of array in order to verify that resizing works correctly.

Comment (ctapobep) [reply src/main/java/hashtable/HashTable2.java:49]: In real projects you wouldn't test internal behaviour of classes. It's important that its interface (public API) works, and how it works - it's not what tests usually check.

File: src/main/java/hashtable/HashTable2.java (hunk 9 above)
Comment (gorshkovayulia) [reply src/main/java/hashtable/HashTable2.java:65]: Hmm...by equals(), I guess

**Pull request merged**